cleaning up system logic + got installed games showing on mac

urwrstkn8mare opened the pull request:
Only been tested on mac so windows logic remains largely the same just reorganised. Complete overhaul of Mac logic as at least on my mac steam's `registry.vdf` does not show an app states but `content_log.txt` is a pretty good alternative. Consolidated all system logic and kept all common logic in `base.py` with all os dependant logic in seperate files.

Note: currently this works better then even master on my mac (as states actually work) so if you'd like to start pushing this to become the latest release i could work on cleaning up the history for `better_protobuf_deps`. otherwise i think just merging it normally should be fine. it won't be pretty but maybe in the future we could work on smaller PRs so its easier to either squash or rebase

ABaumher commented:
tagging in @don-de-marco so i can get a second set of eyes (ngl a lot of the stuff in this branch is not mine anymore so i'm not the best reviewer atm). 

I'll look when i get a chance at merging this in. I'm hesitant to make this the global release because while it addresses a lot of the issues people have been having, the current version is stable for *most* users and this *may* have hidden bugs (i doubt it, our testers are annoyingly good at finding them) That said, i really should because the more people trying to break this the better. 

In the meantime, lets see if i can get @bgebhardt @Mr-Spock1 to test your PR and see if they have similarly improved experiences.

ABaumher commented:
After this merge i'll release this as version 1.2 Alpha and release it on FoG's issue board and see if anyone complains if/when it breaks lol. 

I've got the MVC prototyped out for my general rewrite, once i have implementing the cleaner cache sorted out (a larger "container" for all caches, an abstract base class, implementing a run loop, update persistent storage, etc), i can hopefully slap these caches in there and be done 2.0

ABaumher commented on src/local/base.py:
```diff
@@ -0,0 +1,130 @@
+                return int(app_state["BytesDownloaded"])
+
+
+class BaseClient:
```
Consider inheriting ABC.abc (abstract base class) seeing as the MacOS and windows both inherit them. you can then decorate the functions with `@abstractmethod` and simply `pass` instead of `raise NotImplementedError`. We'll get nice(ish) runtime checks if the abstract classes don't override them. Python 3.12 is adding an `@override` annotation which would really help with typechecking but it's not here yet. Another reason i prefer C#/C++/Java, lol.

ABaumher replied:
could include the overrides module, and seeing as i'm hoping to make the caches implement an abstract parent, it might actually be useful, but for now i'm holding off

urwrstkn8mare replied:
yeah i was thinking about both but `raise NotImplementedError` saved a line of code per method (either way a meaningful error will be raised if not overriden). overrides module could be useful but yeah lets see if we can do it without. i'll make sure to change it to use `abc` 👍

ABaumher commented on requirements/app.txt:
```diff
@@ -11,3 +11,6 @@ rsa==4.9
 betterproto==1.2.5
 
+# local for macos (for now)
+file-read-backwards==3.0.0; sys_platform == 'darwin'
```
we've been trying to keep the install size small because gog requires we pack each module we use along with our plugin, so if you can avoid using this, do so. That said, we're only at 3.5 Mb last i saw so this isn't blocking in any way. It's not worth seeking to the end of the file and manually reading lines backward when the module here does that.

urwrstkn8mare replied:
yeah this module is a pretty small one so it should be alr

ABaumher commented on src/plugin.py:
```diff
@@ -37,41 +33,27 @@
-
-def is_windows():
-    return platform.system().lower() == "windows"
+LAUNCH_DEBOUNCE_TIME = 30
```
👍

don-de-marco commented on src/local/mac.py:
```diff
@@ -0,0 +1,121 @@
+                if words[0] == "AppID":
+                    app_id = words[1]
+                    log.debug("Found app_id: " + app_id)
+                    if not yielded.get(app_id):
```
```suggestion
                    if app_id not in yielded:
```

don-de-marco commented on src/local/mac.py:
```diff
@@ -0,0 +1,121 @@
+                            for k,v in STATE_MAPPING.items():
+                                if k in states:
+                                    yield app_id, v
+                                    yielded[app_id] = True
```
If you're only storing booleans in there, you might want to instead just use a set to store the app_ids

don-de-marco commented on src/local/base.py:
```diff
@@ -0,0 +1,134 @@
+import asyncio
+from contextlib import suppress
+import enum
+import glob
+from logging import getLogger
+import os
+import subprocess
+from typing import Iterable, Optional
+import webbrowser
+from attr import dataclass
+from abc import ABC, abstractmethod
+
+from galaxy.api.types import LocalGame
+
+from .shared import load_vdf
```
the imports could use a bit of ordering (if you're using vscode, just run the "organize imports" command from the palette). this also applies to the rest of the files

don-de-marco commented on src/local/base.py:
```diff
@@ -0,0 +1,134 @@
+            proc = await asyncio.create_subprocess_shell(
+                cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
+            )
+            stdout, stderr = await proc.communicate()
```
```suggestion
            await proc.communicate()
```
you're DEVNULLing the streams anyway

don-de-marco commented on src/local/mac.py:
```diff
@@ -0,0 +1,121 @@
+        # TODO: set max length of log to read
+
+        yielded = {}
+        for i, line in enumerate(lines):
+            if i == MAX_PARSE:
+                break
```
```suggestion
        for i, line in enumerate(lines[:MAX_PARSE]):
```

urwrstkn8mare replied:
the input for lines is usually an iterable so I can't do that

don-de-marco commented on src/local/shared.py:
```diff
@@ -0,0 +1,43 @@
+
+# VDF stuff
+
+class CaseInsensitiveDict(dict):
```
This one is missing an implementation for `get(key, default=None)`. Maybe do a safety-safe there and add it despite you're currently not using it yourself

urwrstkn8mare replied:
well this class is just copy and pasted for earlier. honestly i dont want to go down the rabbit hole of implementing everything for a dict alternative. if its ever needed for more then it can be added does that sound alr?

don-de-marco commented on src/local/shared.py:
```diff
@@ -0,0 +1,43 @@
+def load_vdf(path: str) -> Dict[str, Any]:
+    try:
+        with open(path, encoding="utf-8", errors="replace") as f:
+            yield vdf.load(f, mapper=CaseInsensitiveDict)
```
```suggestion
            return vdf.load(f, mapper=CaseInsensitiveDict)
```

urwrstkn8mare replied:
oh this was actually so that failed parsing values in the with block load_vdf would be used in could caught and logged instead of having to write try catch repeatedly

don-de-marco commented on src/plugin.py:
```diff
@@ -228,79 +213,43 @@ async def _push_cache(self):
-            and self._update_local_games_task.done()
-            and self._regmon.is_updated()
-        ):
+        # TODO: -> local
```
is this still TODO?

urwrstkn8mare replied:
no lol sry

don-de-marco commented on src/plugin.py:
```diff
@@ -228,79 +213,43 @@ async def _push_cache(self):
-            return None
+    # TODO: -> local
+    async def get_local_size(self, game_id: str, context: Dict[str, Manifest]) -> Optional[int]:
+        m = context.get(game_id)
+        if m:
+            return m.app_size()
```
```suggestion
        m = context.get(game_id)
        if not m:
            return None
        return m.app_size()
```

urwrstkn8mare replied:
it does the same thing?

don-de-marco reviewed (commented):
While I'm astonished that parsing the log file seems to be significantly faster than parsing the vdf files, I have one major concern: can you be certain that the log file _always_ contains information about _all_ locally installed games (e.g. is truncation something we need to consider a possibility)? What if you uninstall a game, does it still appear in the log file? (I'd assume so because it's a log and not a state file)

Minor style nitpicks:
* please have a look at the lines where you log information and check the log level at which you log (e.g. there are a few info messages which should rather be debug messages imo)
* please sort import statements at least alphabetically by the respective package name (or, preferably, in accordance to pep8 - there are tools which can assist in that, let me know if I can help find a fit for you)

don-de-marco commented:
I'll give this one a test throughout the evening and see if the Windows part still does it's job. I'll keep you posted

don-de-marco commented:
I have an honest question because I really don't understand it. Why did this not break on Mac, too? The code is identical except for Regmon vs ContentLog in the constructor
```
2023-06-26 18:01:28,789 - galaxy.api.plugin - ERROR - Error while running plugin
Traceback (most recent call last):
  File "C:\Users\XXX\AppData\Local\GOG.com\Galaxy\plugins\installed\steam_ca27391f-2675-49b1-92c0-896d43afa4f8\galaxy\api\plugin.py", line 1142, in create_and_run_plugin
    asyncio.run(coroutine())
  File "D:\obj\Windows-Release\37win32_Release\msi_python\zip_win32\runners.py", line 43, in run
  File "D:\obj\Windows-Release\37win32_Release\msi_python\zip_win32\base_events.py", line 584, in run_until_complete
  File "C:\Users\XXX\AppData\Local\GOG.com\Galaxy\plugins\installed\steam_ca27391f-2675-49b1-92c0-896d43afa4f8\galaxy\api\plugin.py", line 1129, in coroutine
    async with plugin_class(reader, writer, token) as plugin:
  File "C:\Users\XXX\AppData\Local\GOG.com\Galaxy\plugins\installed\steam_ca27391f-2675-49b1-92c0-896d43afa4f8\plugin.py", line 74, in __init__
    self.local = LocalClient()
TypeError: Can't instantiate abstract class WinClient with abstract methods is_updated
```

urwrstkn8mare commented:
hmm that seems to do with changing the baseclass to an abstract base class and the `is_updated` method which is set as a property which seems to trip up `ABC`. 

also i accidentally left the logs in i'll take those out

urwrstkn8mare commented:
> can you be certain that the log file always contains information about all locally installed games

yes but i actually limit how much of the log file i'll read anyways. cause at the end of the day someone could just clear their logs. for state changes reading the log is more than enough but for getting installed games i just get all the manifests and then put the states read from the log on top

urwrstkn8mare commented:
thanks for trying it out i've addressed basically all of your comments and fixed the issue. if you can please try it out again on windows and mac and lmk if it looks good to merge

don-de-marco commented:
lgtm

Please [have a second look at this](https://github.com/ABaumher/galaxy-integration-steam/pull/35/commits/ad6d470ae33e8c183cd5b215eb4a679a6dd704cd#diff-170f694e4968c51471c9828816ec5d8b92125af450a8d5e53ce8f9c036ba74c4L83-R79). I think you by accident inverted the condition which may have cause some unexpected behavior.

ABaumher commented:
> yes but the last state entry is Uninstalled so its even better. 

That's why you read file backwards. That makes sense now. 

I'll let Don-de-marco give it a test tonight and see if it passes. Then I'll see to merging it

urwrstkn8mare commented:
> Please [have a second look at this](https://github.com/ABaumher/galaxy-integration-steam/pull/35/commits/ad6d470ae33e8c183cd5b215eb4a679a6dd704cd#diff-170f694e4968c51471c9828816ec5d8b92125af450a8d5e53ce8f9c036ba74c4L83-R79). I think you by accident inverted the condition which may cause some unexpected behavior.

Thanks for the heads up

don-de-marco reviewed (approved):
I don't see any exceptions in the log and everything seems to be synced so I give this a go.

One heads up for the future: Steam writes a similar log file on Windows so we might want to adapt the changes to Windows as well. In another PR, that is ;)

UncleGoogle reviewed (commented):
@urwrstkn8mare would be good to resurrect tests for local stuff. Those tests were independent from the changed authentiction. It is good to have tests when you do refactor, at least for windows part which is not changed(?).

You can test once or twice manually, but can you do that all the time for growing number of edge cases?

don-de-marco commented:
@UncleGoogle I have made the tests work on another merge request. However, that's a whole different mess to sort out and needs more work - feel free to lend a hand.

ABaumher commented:
Sorry, I'll get to merging this in shortly. Different got users for work and this, has been annoying. So has switching gears from c# to python. Git credential manager made this easier with their latest release... three days ago.

Rebasing this onto master is on the list of things to do. I can't recall if i merged in master already (instead of the rebase) but if i did, mb.

Not enough hours in the day, lol

urwrstkn8mare commented:
> Rebasing this onto master is on the list of things to do

oh @ABaumher  this wasn't built on `master` it was built on `better_proto_deps`

> resurrect tests for local stuff

😱 @UncleGoogle ur right but

UncleGoogle commented:
I just run tests for local games on better_proto_deps and it works (after removing python paths config from pytest.ini).
Imo worth to adjust them till they work

![image](https://github.com/ABaumher/galaxy-integration-steam/assets/12197825/d3b708c2-94a7-4edb-bdea-f0cc6709c17b)

don-de-marco commented:
Well, as I said, I've already been working on getting this to work in #16 including removing obsolete tests and fixing the ones that are still relevant. I suspended the work until we (or more specifically Abaumher) finish getting the code base cleaned up and restructured. There's too much building work going on to keep the test up to par with everything. Also, a significant amount of tests doesn't work anymore since we completely overhauled the protobuf stuff as well as the authentication flow and the tests need to be adapted to the changes. I believe the local games stuff is one of the few tests which still work because we haven't (yet) touched that code

ABaumher commented:
I'm currently testing each function the rework code as i write it, though admittedly i have no idea how to mock asyncio calls. As a workaround, I've just made the calls synchronous and tested with expected data. I'll leave the rest data in as comments. I'm assuming once the code is stable enough these "test" data objects can be mocked in a more concrete manner. 

An issue we're running into is the test cases don't reflect real-world use. Some users have massive libraries or a ton of friends, etc, and the unit tests don't properly catch the issues that crop up as a result. These issues were there in FoG's code but it passed all the tests. Before the work @don-de-marco did (and is still doing), large libraries would consistently kill the plugin not due to any errors in the code but instead because it was too slow or too synchronous/blocking

ABaumher commented:
A counterpoint, of course, is that proper trading would have caught the no 2FA bug that was fixed in 1.0.7. I guess the moral of the story is i need to finish the rework faster so we can properly test before releasing 2.0

don-de-marco commented:
I can't help you there, either. It's probably best if you just look into how the already existing tests do the asyncio stuff. I know for certain that we have the means to _test_ asyncio methods. But I do not know if and how one would mock them.

ABaumher commented:
oh god, now i need to merge all this into my conversion branch and prep all of this for nebula's clean history. can't wait! /s

UncleGoogle commented:
Unit tests wont catch a lot of things, but should find a lot of dummy mistakes (and a lot of time for testing broken build). I cant imagine to dont have running tests suite since I started to practice TDD. The same with the CI running tests. Ill add ci running what we have green already if find a time.
